x-pack/plugins/security_solution/server/endpoint/services/actions/clients/sentinelone/sentinel_one_actions_client.ts

```diff
@@ -249,28 +303,238 @@ export class SentinelOneActionsClient extends ResponseActionsClientImpl {
+          `${command} response action ID [${action.EndpointActions.action_id}] missing SentinelOne agent ID, thus unable to check on it's status. Forcing it to complete as failure.`
+        );
+
+        completedResponses.push(
+          this.buildActionResponseEsDoc<{}, SentinelOneIsolationResponseMeta>({
+            actionId: action.EndpointActions.action_id,
+            agentId: Array.isArray(action.agent.id) ? action.agent.id[0] : action.agent.id,
+            data: { command },
+            error: {
+              message: `Unable to very if action completed. SentinelOne agent id ('meta.agentId') missing on action request document!`,
+            },
+          })
+        );
+      }
+
+      return acc;
+    }, [] as QueryDslQueryContainer[]);
+
+    if (agentListQuery.length > 0) {
+      const query: QueryDslQueryContainer = {
+        bool: {
+          must: [
+            {
+              terms: {
+                // Activity Types can be retrieved from S1 via API: `/web/api/v2.1/activities/types`
+                'sentinel_one.activity.type':
+                  command === 'isolate'
+                    ? [
+                        // {
+                        //    "id": 1001
+                        //    "action": "Agent Disconnected From Network",
+                        //    "descriptionTemplate": "Agent {{ computer_name }} was disconnected from network.",
+                        // },
+                        1001,
+
+                        // {
+                        //    "id": 2010
+                        //    "action": "Agent Mitigation Report Quarantine Network Failed",
+                        //    "descriptionTemplate": "Agent {{ computer_name }} was unable to disconnect from network.",
+                        // },
+                        2010,
+                      ]
+                    : [
+                        // {
+                        //    "id": 1002
+                        //    "action": "Agent Reconnected To Network",
+                        //    "descriptionTemplate": "Agent {{ computer_name }} was connected to network.",
+                        // },
+                        1002,
+                      ],
+              },
+            },
+          ],
+          should: agentListQuery,
+          minimum_should_match: 1,
+        },
+      };
+
+      const searchRequestOptions: SearchRequest = {
+        index: SENTINEL_ONE_ACTIVITY_INDEX,
+        query,
+        // There may be many documents for each host/agent, so we collapse it and only get back the
+        // first one that came in after the isolate request was sent
+        collapse: {
+          field: 'sentinel_one.activity.agent.id',
+          inner_hits: {
+            name: 'first_found',
+            size: 1,
+            sort: [{ 'sentinel_one.activity.updated_at': 'asc' }],
+          },
+        },
+        // we don't need the source. The document will be stored in `inner_hits.first_found`
```

Comment:
👍  I've heard `fields` is preferred above `_source`  when possible. Does `collapse` work the same way as fields meaning it returns just the `agent.id` field, or the whole document?

Reply:
Not sure. Collapse (in my understanding) takes the search results, groups them by a given field and then applies additional logic to it. Son in this case, it groups them by S1 agent ID, then than it sorts those in ascending order, and then it takes the first (most recent) record from that and it stores it in a property called `first_found`. 

You can see what `collapse` does by running this in the dev console (just update the Agent ids below):

```
GET logs-sentinel_one.activity-default/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "terms": {
            "sentinel_one.activity.type": [
              1001,
              2010
            ]
          }
        }
      ],
      
      "should": [
        {
          "bool": {
            "filter": [
              { "term": { "sentinel_one.activity.agent.id": "1913920934584665209" } },
              { "range": { "sentinel_one.activity.updated_at": { "gt": "2024-04-02T12:59:22.229Z" } } }
            ]
          }
        }
      ],
      "minimum_should_match": 1
    }
  },
  "_source": false,
  "collapse": {
    "field": "sentinel_one.activity.agent.id",
    "inner_hits": {
      "name": "first_found",
      "size": 1,
      "sort": [ { "sentinel_one.activity.updated_at": "asc" } ]
    }
  }, 
  "size": 1000,
  "sort": [
    {
      "sentinel_one.activity.updated_at": {
        "order": "desc"
      }
    }
  ]
}
```

you can also reference the docs here: https://www.elastic.co/guide/en/elasticsearch/reference/current/collapse-search-results.html